### PR TITLE
Add dropdown for selecting publisher logo

### LINKED ISSUE #5108: Editor: Cannot use multiple publisher logos
## Bug Description

It seems like the publisher logo section in the Document Panel does not work as expected. Maybe the spec is unclear or we never fully implemented this.

[The spec currently says](https://docs.google.com/document/d/1GJZljicdXcP7oRdywsF3EGUaCVpCHsDrK2gUMubnrbk/edit?pli=1#heading=h.ymf85pwospli):

> Allows the author to pick a publisher logo, as defined in AMP Stories’ metadata guidelines.
> 
> Inherited from global plugin settings, but can be overridden for the story at hand locally.

**[UX / Design specs](https://www.figma.com/file/Cd9iacVknHK62RO4pRwb68/Stories-Editor?node-id=5837%3A4)**

## Expected Behaviour

I should be able to use different publisher logos for different stories.

## Steps to Reproduce

1. Create story 1, select publisher logo A.
1. Publish story 1.
1. See how logo A is selected as the default in the dashboard settings (so far so good)
1. Create story 2, select publisher logo B.
1. Publish story 2.
1. See how logo B is selected as the default in the dashboard settings (so far so good)
1. Edit story 1
1. See logo B in the document panel instead of A (note: the saved story still contains logo A)
1. Save story 1, now the saved markup contains logo B

## Additional Context

- Plugin Version: 1.1.0
- Operating System: macOS
- Browser: Chrome 86;

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by merapi**
## Summary

Instead of upload image dialog, this adds a dropdown list to select the publisher logo for a given story (each story has its own publisher logo).
According to https://xwp.slack.com/archives/C01ESBB8ZN1/p1617715788058500?thread_ts=1617197455.025200&cid=C01ESBB8ZN1

Before:
image

Now:
| Empty state, no logos uploaded | Default logo is selected when available | Dropdown expanded |
|---|---|---|
| Screenshot 2021-09-09 at 16 55 40 | Screenshot 2021-09-09 at 17 01 53 | Screenshot 2021-09-09 at 17 02 55 |
|---|---|---|

## Relevant Technical Choices

- I've added an API function for fetching the settings.

## To-do

- [ ] File a bug about dropdown: when search is disabled, navigation with keyboard (arrows) is broken
- [ ] File a bug about dropdown: dropdown cannot be highlighted (from the pre-publish checklist)

## User-facing changes

Well explained above.

## Testing Instructions

I've tested this code with:

1. No publisher logos (hard to do on QA env because we cannot remove the last logo?)
2. Saved the story with no publisher logos then added a logo in settings and checked the story - because it was saved without the logo it stays like that (is that ok?)
3. Tested that each story has its own logo and uploading a new one doesn't change any of them.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #5108

### CODE REVIEW

File: packages/story-editor/src/components/panels/document/publish/publish.js
Comment (merapi) [inline packages/story-editor/src/components/panels/document/publish/publish.js]: Should it open in a new tab or do we want to rely on the `UnsavedChangesWarning`? also, is there a better way than hardcoding the URL like this?

**Comment by swissspidy** [comment]
This PR seems to remove the ability to upload a custom publisher logo to the editor [as described in Figma](https://www.figma.com/file/Cd9iacVknHK62RO4pRwb68/Stories-Editor?node-id=5837%3A4).

Screenshot 2021-09-09 at 17 44 57

Removing the ability to upload a custom publisher logo is quite the step backward.

**Comment by spacedmonkey** [comment]
> Removing the ability to upload a custom publisher logo is quite the step backward. 

Not to mention, removing all the work I did lately into cropping the image when it is picked.

**Comment by merapi** [comment]
> Removing the ability to upload a custom publisher logo is quite the step backward.

You can still upload them on the Settings page.

We need a product decision because I feel like [Omar iterated over what we currently have](https://xwp.slack.com/archives/C01ESBB8ZN1/p1617197455025200) vs developing new components as per Figma (which are not ideal since as soon somebody will upload the first logo there is zero discoverability in that DND feature [other parts of the editor don't have it, like poster image next to it, so it's a bit confusing when you can do that]).

From the Slack conversation, [it was clear](https://xwp.slack.com/archives/C01ESBB8ZN1/p1617717591059000?thread_ts=1617197455.025200&cid=C01ESBB8ZN1) that upload will be only possible via Manage link, and no one argued against it.

**Comment by merapi** [comment]
> Not to mention, removing all the work I did lately into cropping the image when it is picked.

BTW: shouldn't that be also possible on the Settings page?

**Comment by swissspidy** [comment]
> > Removing the ability to upload a custom publisher logo is quite the step backward.
> 
> You can still upload them on the Settings page.
> 
> We need a product decision because I feel like [Omar iterated over what we currently have](https://xwp.slack.com/archives/C01ESBB8ZN1/p1617197455025200) vs developing new components as per Figma (which are not ideal since as soon somebody will upload the first logo there is zero discoverability in that DND feature [other parts of the editor don't have it, like poster image next to it, so it's a bit confusing when you can do that]).

@choumx @aaskedall Can we revisit this please and?

There should be a way to upload a new publisher logo directly from within the editor & crop it. Doesn't have to be via drag and drop, but at least there should be an upload button like we currently have.

> From the Slack conversation, [it was clear](https://xwp.slack.com/archives/C01ESBB8ZN1/p1617717591059000?thread_ts=1617197455.025200&cid=C01ESBB8ZN1) that upload will be only possible via Manage link, and no one argued against it.

At least to me that's not very clear at all and not how I interpret the discussion there.

> > Not to mention, removing all the work I did lately into cropping the image when it is picked.
> 
> BTW: shouldn't that be also possible on the Settings page?

We're tracking this in #4148.

**Comment by merapi** [comment]
> At least to me that's not very clear at all and not how I interpret the discussion there.

_"click manage to do **anything** else (e.g., change the default, etc.)"_ - was not super clear to me either, but 
_"Yep, that’s how I understand Figma as well — the user can either choose between the existing directly or click “Manage”_ to **upload new** / set default." and Omar's +1 removed any doubt + it was not a big deal for me (how often do you upload a new publisher logo?), it's still just a few clicks away.

**Comment by merapi** [comment]
Some options for us (both open media picker with crop + we need to make an additional API call to add the selected image to publisher logos, right?):

pencil icon somewhere:
image

as one of the options (should be first? so you don't have to scroll):
image

**Comment by dreamofabear** [comment]
Good catch and discussion. That Slack thread was back in March (before we added the ability to crop publisher logo on upload in June/July #7744), so it's understandable there's some confusion.

I agree with Pascal that we should keep current functionality (upload custom publisher logo and crop it) while adding a dropdown to select from already-uploaded logos. 

I like Marcin's option 2:
> as one of the options (should be first? so you don't have to scroll):
> image

Can we add or just use an icon for the "Open media picker" button that's nested inside the dropdown? Maybe reuse the same upload ☁️ ⬆️   icon that we have for the media library. @aaskedall for input.

**Comment by merapi** [comment]
Just the icon:
| image  | image  |
|---|---|

It looks like one of the logos 🤔 

Another option (BTW: it's not only upload, because it opens media picker):
image

**Comment by spacedmonkey** [comment]
>dropdown to select from already-uploaded logos.

This is no longer possible as of https://github.com/google/web-stories-wp/pull/8946.  In this line we remove the updating of [option](https://github.com/google/web-stories-wp/pull/8946/files#diff-6b011397c22cc9aff84c1cf303869ae7444ff773dcbaaeef8b61e5d0520fc326L171). 

We should keep track of logos a user has uploaded before. So if a user creates 3 stories, all with different logos and goes to make a 4th, then the should be about to select any of the publisher logos they used before.

**Comment by swissspidy** [comment]
> > dropdown to select from already-uploaded logos.
> 
> This is no longer possible as of #8946. In this line we remove the updating of [option](https://github.com/google/web-stories-wp/pull/8946/files#diff-6b011397c22cc9aff84c1cf303869ae7444ff773dcbaaeef8b61e5d0520fc326L171).
> 
> We should keep track of logos a user has uploaded before. So if a user creates 3 stories, all with different logos and goes to make a 4th, then the should be about to select any of the publisher logos they used before.

That's as simple as getting all existing logos via `GET /wp-json/web-stories/v1` and updating `web_stories_publisher_logos` via `POST /wp-json/web-stories/v1` after uploading accordingly.

**Comment by merapi** [comment]
> This is no longer possible as of #8946. In this line we remove the updating of [option](https://github.com/google/web-stories-wp/pull/8946/files#diff-6b011397c22cc9aff84c1cf303869ae7444ff773dcbaaeef8b61e5d0520fc326L171).

You can still upload logos on the Settings page (even without what we will do next here).

> We should keep track of logos a user has uploaded before. So if a user creates 3 stories, all with different logos and goes to make a 4th, then the should be about to select any of the publisher logos they used before.

Yep, as I said:
> Some options for us (both open media picker with crop + we need to make an additional API call to add the selected image to publisher logos, right?):

**Comment by spacedmonkey** [comment]
> That's as simple as getting all existing logos via GET /wp-json/web-stories/v1 and updating web_stories_publisher_logos via POST /wp-json/web-stories/v1 after uploading accordingly.

I guess you mean, updating settings endpoint. Sadly editors and authors can not do this, as they do not have the manage options permission.

**Comment by swissspidy** [comment]
Good point. We should not offer these options for users without the necessary capabilities. Just like today these users can't upload publisher logos already if they cannot upload media.

**Comment by merapi** [comment]
> Good point. We should not offer these options for users without the necessary capabilities. Just like today these users can't upload publisher logos already if they cannot upload media.

I'm already checking for `hasUploadMediaAction` when showing that extra button, `canManageSettings` will be used for the link, but, we need to solve one more issue: those roles can't even read the settings to get the publisher logos list - extra endpoint or get it when fetching the story data? otherwise, they can't change the logo, they must use the default one (if set).

**Comment by spacedmonkey** [comment]
I would hook into the publisher logo id being saved in post meta, get the existing active publisher logos and if the id is not in the list, update the option. This should be done in the backend for the user, no rest api request needed. 

I can prototype something, if this is at all helpful

**Comment by merapi** [comment]
🤔   so how do we do this (as per spec/Figma): _"drop-down where you select the publisher logo, and it would show you all the configured publisher logos"_

**Comment by swissspidy** [comment]
Folks, let's create a ticket for this. We can discuss specifics there.

> I would hook into the publisher logo id being saved in post meta, get the existing active publisher logos and if the id is not in the list, update the option. This should be done in the backend for the user, no rest api request needed.

That does not address the part where users need to get a list of all publisher logos.

> I'm already checking for `hasUploadMediaAction` when showing that extra button, `canManageSettings` will be used for the link, but, we need to solve one more issue: those roles can't even read the settings to get the publisher logos list - extra endpoint or get it when fetching the story data? 

Yeah an extra endpoint sounds reasonable to me for this.

> otherwise, they can't change the logo, they must use the default one (if set).

Right. Let's go with this for now until we have something implemented for you.

**Comment by merapi** [comment]
> Let's go with this for now until we have something implemented for you.

Should be ready for review now:
image

**Comment by miina** [comment]
@merapi Some tests are failing FYI.

File: packages/story-editor/src/components/panels/document/publish/publish.js
Comment (miina) [inline packages/story-editor/src/components/panels/document/publish/publish.js]: There's `dashboardSettingsLink` in config that can be used here.

File: packages/story-editor/src/components/panels/document/publish/publish.js
Comment (miina) [inline packages/story-editor/src/components/panels/document/publish/publish.js]: Maybe for readability we could do something like this:
```
const displayText = publisherLogos.length ? __( 'Select logo', ... ) : __( 'No logo', ... );
const activeItemRenderer = () => publisherLogo.id ? <LogoImg .... /> : <Text ...>{displayText}</Text>;
```

File: packages/story-editor/src/components/panels/document/publish/publish.js
Comment (miina) [inline packages/story-editor/src/components/panels/document/publish/publish.js]: This looks like something that could be turned into `getErrorMessage`, accepting param to return the message either for the poster or publisher (unless we move the publisher logo logic to a separate file).

File: packages/story-editor/src/components/panels/document/publish/publish.js
Comment (miina) [inline packages/story-editor/src/components/panels/document/publish/publish.js:323]: Maybe the Publisher logo would deserve its own file now?

File: packages/story-editor/src/components/form/advancedDropDown/container.js
Comment (miina) [inline packages/story-editor/src/components/form/advancedDropDown/container.js:43]: Q: Does this change influence any other use cases of the advanced dropdown?

File: packages/story-editor/src/app/story/effects/useLoadStory.js
Comment (miina) [inline packages/story-editor/src/app/story/effects/useLoadStory.js]: Maybe extract `embedded?.['wp:publisherlogo']?.[0]` into a separate constant for readability?

**Review by miina — COMMENTED** [review]
Works great generally.

One issue from testing.
When choosing new Media, after cropping, the Media Library does not close (even though the image gets cropped and selected):
cropping

Let me know if that's a bug that existed before already, will approve then.

File: packages/story-editor/src/components/panels/document/publish/publish.js
Comment (miina) [inline packages/story-editor/src/components/panels/document/publish/publish.js]: Ah, okay, now I see where the Media Picker "not closing" bug comes from. `onClick={openMediaPicker}` is used twice here -- once on the option and once on `Text`. So it gets opened twice.

**Comment by swissspidy** [comment]
@merapi @spacedmonkey Created #9007 to address the open questions from the above discussion. PTAL

File: packages/story-editor/src/components/panels/document/publish/publish.js
Comment (merapi) [reply packages/story-editor/src/components/panels/document/publish/publish.js]: Yep, thanks for catching that.

File: packages/story-editor/src/components/form/advancedDropDown/container.js
Comment (merapi) [reply packages/story-editor/src/components/form/advancedDropDown/container.js:43]: We use it for font picker and authors, both are safe (they are 100% width, 276px most of the time).

File: packages/story-editor/src/app/story/effects/useLoadStory.js
Comment (merapi) [reply packages/story-editor/src/app/story/effects/useLoadStory.js]: I was thinking about that but we don't do that generally(?)
image

File: packages/story-editor/src/components/panels/document/publish/publish.js
Comment (merapi) [reply packages/story-editor/src/components/panels/document/publish/publish.js]: Done in https://github.com/google/web-stories-wp/pull/8989/commits/4756907aa4ce2e383463fdc6e37848966af7a60b.
About moving to its own file - no strong opinion, it's just a bit longer than it was before.

Comment (miina) [reply packages/story-editor/src/components/panels/document/publish/publish.js]: Yep, that would be the reason -- not to allow files to grow very large whenever possible. Let's keep it as is for now then and do it whenever it should grow more.

File: packages/story-editor/src/app/story/effects/useLoadStory.js
Comment (miina) [reply packages/story-editor/src/app/story/effects/useLoadStory.js]: Hm, okay. Not sure what's the reason 🤔

**Review by miina — APPROVED** [review]
LGTM Pending tests.

File: packages/e2e-tests/src/specs/editor/inspector.js
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/inspector.js]: Why is this now skipped?

Comment (merapi) [reply packages/e2e-tests/src/specs/editor/inspector.js]: I'm just testing what's going on here (_"Test fail reason"_).
Locally I'm getting this, any ideas?

image

Comment (swissspidy) [reply packages/e2e-tests/src/specs/editor/inspector.js]: The easiest way to find out is by opening the editor as a contributor user and then checking the browser console.
The more complicated way would be by running your tests in non-headless mode and observing the behavior.

Currently, they contributors get a 403 because they don't have access to the `web-stories/v1/settings` route.

You need to check for `canManageSettings` capabilities before trying to access it.

As per our lengthy discussion above, I opened #9007 to move publisher logos to a separate REST route that's open for contributors. I might get to that tomorrow if Jonny doesn't before me.

Until then, you'd need to work around it.

Comment (swissspidy) [reply packages/e2e-tests/src/specs/editor/inspector.js]: Happy to chat about it tomorrow if things are unclear.

**Comment by swissspidy** [comment]
@merapi since this PR is still WIP, now it would be possible to leverage the new publisher logos REST API routes here.

**Comment by merapi** [comment]
> @merapi since this PR is still WIP, now it would be possible to leverage the new publisher logos REST API routes here.

After checking what was added/changed I've decided that it will be better to work a little more on this and utilize the new stuff (just resolving conflicts is not an option).

File: packages/story-editor/src/components/panels/document/publish/publish.js
Comment (swissspidy) [inline packages/story-editor/src/components/panels/document/publish/publish.js]: ⚠️ 

This needs to use the `<MediaUpload` component from the config, something like this:

```js
const { MediaUpload } = useConfig();

// ...

<MediaUpload
  onSelect={onNewPublisherLogoSelected}
  onSelectErrorMessage={publisherLogoErrorMessage}
  type={allowedImageMimeTypes}
  render={renderUploadButtonIcon}
title={__('Select as publisher logo', 'web-stories')}
/>
```

File: packages/story-editor/package.json
Comment (swissspidy) [inline packages/story-editor/package.json]: ⚠️  `wp-story-editor` requires `story-editor`, not the other way around.

See my other comment about how not to use `useMediaPicker`

File: packages/story-editor/src/components/panels/document/publish/publish.js
Comment (swissspidy) [inline packages/story-editor/src/components/panels/document/publish/publish.js]: This needs to be removes as per my other comment

File: packages/wp-story-editor/src/api/media.js
Comment (swissspidy) [inline packages/wp-story-editor/src/api/media.js]: Is this still needed? `getPublisherLogos` returns all publisher logos + their URLs

Comment (merapi) [reply packages/wp-story-editor/src/api/media.js]: Nah, I just didn't revert that.

File: packages/story-editor/src/components/panels/document/publish/publish.js
Comment (miina) [inline packages/story-editor/src/components/panels/document/publish/publish.js:246]: Isn't 32x32 the default for icons btw?

Comment (merapi) [reply packages/story-editor/src/components/panels/document/publish/publish.js:246]: It will still grow depending on the context, without it: 
image

**Review by swissspidy — APPROVED** [review]
Seems to work well 👍 🚀

**Review by spacedmonkey — DISMISSED** [review]
I am seeing an error when I try and set as an author. 

https://user-images.githubusercontent.com/237508/135874128-8f114b64-bc45-40e8-a6f6-d4e74d131bec.mov

CC @swissspidy

**Comment by swissspidy** [comment]
@spacedmonkey Image cropping for authors also appears to be broken on `main` and is not related to this PR. So this warrants its own ticket.

**Comment by swissspidy** [comment]
@merapi The `should show cards that require the hasUploadMediaAction permission` karma test appears to be failing in this PR.

**Comment by swissspidy** [comment]
> @spacedmonkey Image cropping for authors also appears to be broken on `main` and is not related to this PR. So this warrants its own ticket.

Looks like it's because authors can only edit/crop attachments they have uploaded themselves, but not ones uploaded by other users.

**Review by spacedmonkey — APPROVED** [review]

**Comment by miina** [comment]
> The should show cards that require the hasUploadMediaAction permission karma test appears to be failing in this PR.

@swissspidy Looks like choosing the poster image was broken -- not sure if it came from merge issues or something else but I changed it back to how it is in `main` and now it seems to work both in the editor and for tests (locally at least, waiting for the tests to run here, too): c5c1a79

**Pull request merged**
